lib/fastlane/plugin/wpmreleasetoolkit/helper/git_helper.rb

```diff
@@ -5,12 +5,42 @@ module Helper
     # Helper methods to execute git-related operations
     #
     module GitHelper
-      # Checks if the current directory is (inside) a git repo
+      # Checks if the given path, or current directory if no path is given, is
+      # inside a Git repository
       #
-      # @return [Bool] True if the current directory is the root of a git repo (i.e. a local working copy) or a subdirectory of one.
+      # @param [String] path An optional path where to check if a Git repo
+      #        exists.
       #
-      def self.is_git_repo?
-        system 'git rev-parse --git-dir 1> /dev/null 2>/dev/null'
+      # @return [Bool] True if the current directory is the root of a git repo
+      #         (i.e. a local working copy) or a subdirectory of one.
+      #
+      def self.is_git_repo?(path: nil)
+        args = ['git']
+
+        unless path.nil?
+          # `path` might represent a file that doesn't yet exist but is in a valid Git
+          # repo. In such case, `git rev-parse`, which we use for the repo
+          # check, would fail. To avoid that, let's perform the check on the first existing ancestor of the given path (which will be the path itself if it exists)
+          # one is part of a Git repository.
+          path = first_existing_ancestor_of(path: path)
+          args += ['-C', File.file?(path) ? File.dirname(path) : path]
+        end
+
+        args += ['rev-parse']
+
+        Action.sh(args, print_command_output: false) do |status, _, _|
+          status.success?
+        end
```

Comment:
Unfortunately, because of [how `sh` works internally](https://github.com/fastlane/fastlane/blob/7ba2d6a31e47c7919b3847fdbbde72feec68d213/fastlane/lib/fastlane/helper/sh_helper.rb#L67-L81) this setup will result in an error message in the Fastlane logs in case of a valid path that doesn't exist yet:

![image](https://user-images.githubusercontent.com/1218433/117906403-2d910d80-b318-11eb-80d7-a714e0ca15ab.png)

I'm wondering whether I should add a `UI.message` of sort to let the users know that's an expected error? Or maybe go through the longer process of adding an option for suppressing that output in Fastlane?

Reply:
We use [this approach](https://github.com/wordpress-mobile/release-toolkit/blob/c476c50f507093ed30070740803083a4d4a42b49/lib/fastlane/plugin/wpmreleasetoolkit/helper/filesystem_helper.rb#L10-L30) to determining the root of a git repo elsewhere, I wonder if we could use something along those lines instead of relying on `git` to return specific error codes?

WDYT?

Reply:
Thanks @jkmassel 🙇 

One thing to point out first:

> instead of relying on `git` to return specific error codes?

This implementation relies on `git rev-parse` succeeding or failing, it does not inspect the error code. I think this is an important difference because it makes the logic simpler, it either succeeded or not, no need for extra info. Looking at the error code was one of the first things I tried (it was the error message, actually) but I wasn't happy with it either. Thanks to @AliSoftware's suggestions, I was able to get to this more flexible "did it succeed?" state.

Having said that, you make a legit point anyways. I'm of two minds on whether it's best to delegate to Git or implement the logic ourselves...

It's fair to say that we can trust Git to know how to do this job, and it would save us from reinventing the wheel. At the same time, shelling out is never nice and, like in this instance, we can run into awkward behavior regarding the output of the shelled-out command.

🤔 

I consider this an implementation detail, so I'm going to give it a go in a different branch. In the best case scenario, we like what I come up with and we merge it in here. If instead it may require extra work, we could still merge this working implementation and iterate on that approach async.